File: android-app/app/src/main/kotlin/com/mysplitwise/app/model/Expense.kt
package com.mysplitwise.app.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Per-user breakdown of a single expense. sum(paid)===amount, sum(owed)===amount. */
@Serializable
data class ExpenseShare(
    val userId: String,
    val paid: Double,
    val owed: Double,
)

@Serializable
data class Comment(
    val id: String,
    val userId: String,
    val text: String,
    val createdAt: String,
)

@Serializable
data class LineItem(
    val id: String,
    val name: String,
    val amount: Double,
    val participantIds: List<String>,
)

@Serializable
enum class PaymentMethod {
    @SerialName("cash") CASH,
    @SerialName("card") CARD,
    @SerialName("wire") WIRE,
    @SerialName("zelle") ZELLE,
    @SerialName("venmo") VENMO,
    @SerialName("cashapp") CASHAPP,
    @SerialName("paypal") PAYPAL,
    @SerialName("other") OTHER,
}

/** Port of the `Expense` interface in `src/lib/types.ts`. */
@Serializable
data class Expense(
    val id: String,
    val description: String,
    val amount: Double,
    val currency: String,
    val category: String,
    /** ISO date string, e.g. "2026-03-14". */
    val date: String,
    val groupId: String? = null,
    val shares: List<ExpenseShare>,
    val createdBy: String,
    val createdAt: String,
    val isSettlement: Boolean,
    val notes: String? = null,
    val receiptUrl: String? = null,
    val comments: List<Comment>? = null,
    val recurringId: String? = null,
    val items: List<LineItem>? = null,
    val tax: Double? = null,
    val tip: Double? = null,
    val paymentMethod: PaymentMethod? = null,
    /** Settlement dispute audit trail — visibility only, not fraud prevention. */
    val disputed: Boolean? = null,
    val disputeReason: String? = null,
    val disputedBy: String? = null,
    val disputedAt: String? = null,
)

File: ios-app/MysplitwiseCore/Sources/MysplitwiseCore/Models/Expense.swift
import Foundation

/// Per-user breakdown of a single expense. sum(paid)===amount, sum(owed)===amount.
public struct ExpenseShare: Codable, Equatable, Hashable {
    public var userId: String
    public var paid: Double
    public var owed: Double

    public init(userId: String, paid: Double, owed: Double) {
        self.userId = userId
        self.paid = paid
        self.owed = owed
    }
}

public struct Comment: Codable, Identifiable, Equatable, Hashable {
    public var id: String
    public var userId: String
    public var text: String
    public var createdAt: String

    public init(id: String, userId: String, text: String, createdAt: String) {
        self.id = id
        self.userId = userId
        self.text = text
        self.createdAt = createdAt
    }
}

public struct LineItem: Codable, Identifiable, Equatable, Hashable {
    public var id: String
    public var name: String
    public var amount: Double
    public var participantIds: [String]

    public init(id: String, name: String, amount: Double, participantIds: [String]) {
        self.id = id
        self.name = name
        self.amount = amount
        self.participantIds = participantIds
    }
}

public enum PaymentMethod: String, Codable, CaseIterable {
    case cash, card, wire, zelle, venmo, cashapp, paypal, other
}

/// Port of the `Expense` interface in `src/lib/types.ts`.
public struct Expense: Codable, Identifiable, Equatable, Hashable {
    public var id: String
    public var description: String
    public var amount: Double
    public var currency: String
    public var category: String
    /// ISO date string, e.g. "2026-03-14".
    public var date: String
    public var groupId: String?
    public var shares: [ExpenseShare]
    public var createdBy: String
    public var createdAt: String
    public var isSettlement: Bool
    public var notes: String?
    public var receiptUrl: String?
    public var comments: [Comment]?
    public var recurringId: String?
    public var items: [LineItem]?
    public var tax: Double?
    public var tip: Double?
    public var paymentMethod: PaymentMethod?
    /// Settlement dispute audit trail — visibility only, not fraud prevention.
    public var disputed: Bool?
    public var disputeReason: String?
    public var disputedBy: String?
    public var disputedAt: String?

    public init(
        id: String, description: String, amount: Double, currency: String, category: String,
        date: String, groupId: String?, shares: [ExpenseShare], createdBy: String, createdAt: String,
        isSettlement: Bool, notes: String? = nil, receiptUrl: String? = nil, comments: [Comment]? = nil,
        recurringId: String? = nil, items: [LineItem]? = nil, tax: Double? = nil, tip: Double? = nil,
        paymentMethod: PaymentMethod? = nil, disputed: Bool? = nil, disputeReason: String? = nil,
        disputedBy: String? = nil, disputedAt: String? = nil
    ) {
        self.id = id
        self.description = description
        self.amount = amount
        self.currency = currency
        self.category = category
        self.date = date
        self.groupId = groupId
        self.shares = shares
        self.createdBy = createdBy
        self.createdAt = createdAt
        self.isSettlement = isSettlement
        self.notes = notes
        self.receiptUrl = receiptUrl
        self.comments = comments
        self.recurringId = recurringId
        self.items = items
        self.tax = tax
        self.tip = tip
        self.paymentMethod = paymentMethod
        self.disputed = disputed
        self.disputeReason = disputeReason
        self.disputedBy = disputedBy
        self.disputedAt = disputedAt
    }
}
